File: Python Problems/BiTreeZigZag.py
#  Given the root of a binary tree, return the zigzag level order traversal of its nodes' values. 
#  (i.e., from left to right, then right to left for the next level and alternate between).


def zigzagLevelOrder(root):
    if not root:
        return []
    
    result = []
    queue = [root]
    level = 0
    
    while queue:
        level_nodes = []
        size = len(queue)
        
        for _ in range(size):
            node = queue.pop(0)
            if level % 2 == 0:
                level_nodes.append(node.val)  # Left to right
            else:
                level_nodes.insert(0, node.val)  # Right to left
            
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        
        result.append(level_nodes)
        level += 1
    
    return result

File: JavaScript Problems/biTreeZigZag.js
// Given the root of a binary tree, return the zigzag level order traversal of its nodes' values. 
// (i.e., from left to right, then right to left for the next level and alternate between).


function zigzagLevelOrder(root) {
    if (!root) return [];
    
    const result = [];
    let level = 0;
    
    function traverse(node, level) {
        if (!node) return;
        
        if (!result[level]) {
            result[level] = [];
        }
        
        if (level % 2 === 0) {
            result[level].push(node.val);  // Left to right
        } else {
            result[level].unshift(node.val);  // Right to left
        }
        
        traverse(node.left, level + 1);
        traverse(node.right, level + 1);
    }
    
    traverse(root, level);
    return result;
}
